evaluator/autoware_trajectory_evaluator/src/trajectory_evaluator_node.cpp

```diff
@@ -0,0 +1,234 @@
+      trajectory_with_timestamp.trajectory_points.push_back(point_with_time);
+    }
+
+    calculate_time_from_start(trajectory_with_timestamp, current_position, 1e-3);
+    trajectory_history_.push_back(trajectory_with_timestamp);
+    if (trajectory_history_.size() > traj_history_limit) {
+      trajectory_history_.erase(trajectory_history_.begin());
+    }
+  }
+}
+
+void TrajectoryEvaluatorNode::calculate_time_from_start(
+  TrajectoryWithTimestamp & trajectory, const geometry_msgs::msg::Point & current_ego_point,
+  const float min_velocity)
+{
+  auto & trajectory_points = trajectory.trajectory_points;
+  std::vector<autoware_planning_msgs::msg::TrajectoryPoint> trajectory_points_sequence;
+  for (const auto & traj_point_with_time : trajectory_points) {
+    trajectory_points_sequence.push_back(traj_point_with_time.trajectory_point);
+  }
+
+  const auto nearest_segment_idx =
+    autoware::motion_utils::findNearestSegmentIndex(trajectory_points_sequence, current_ego_point);
+
+  if (nearest_segment_idx + 1 >= trajectory_points.size()) {
+    return;
+  }
+
+  for (auto & point_with_time : trajectory_points) {
+    point_with_time.time_from_start = rclcpp::Duration::from_seconds(0);
+  }
+
+  for (size_t idx = nearest_segment_idx + 1; idx < trajectory_points.size(); ++idx) {
+    const auto & from = trajectory_points[idx - 1];
+    const auto velocity = std::max(min_velocity, from.trajectory_point.longitudinal_velocity_mps);
+
+    if (velocity != 0.0) {
+      auto & to = trajectory_points[idx];
+      const double distance = autoware::universe_utils::calcDistance2d(
+        from.trajectory_point.pose, to.trajectory_point.pose);
+      const double time_delta = distance / velocity;
+      to.time_from_start =
+        rclcpp::Duration::from_seconds(time_delta) + trajectory_points[idx - 1].time_from_start;
+    }
+  }
+}
+
+void TrajectoryEvaluatorNode::on_kinematic_state(
+  const nav_msgs::msg::Odometry::ConstSharedPtr odom_msg)
+{
+  if (odom_msg != nullptr) {
+    const double velocity = std::sqrt(
+      std::pow(odom_msg->twist.twist.linear.x, 2) + std::pow(odom_msg->twist.twist.linear.y, 2) +
+      std::pow(odom_msg->twist.twist.linear.z, 2));
+
+    if (velocity == 0.0) {
+      return;
+    }
+
+    rclcpp::Time current_time = this->now();
+
+    for (const auto & trajectory : trajectory_history_) {
+      double min_distance = std::numeric_limits<double>::max();
+      size_t closest_index = trajectory.trajectory_points.size();
+      const geometry_msgs::msg::Pose & pose = odom_msg->pose.pose;
+      for (size_t i = 0; i < trajectory.trajectory_points.size(); ++i) {
+        const auto & trajectory_point_pose = trajectory.trajectory_points[i].trajectory_point.pose;
+        double distance = autoware::universe_utils::calcDistance2d(trajectory_point_pose, pose);
+
+        if (distance < min_distance) {
+          min_distance = distance;
+          closest_index = i;
+        }
+      }
```

Comment:
There is already a function `autoware::motion_utils::findNearestIndex()` to find the nearest index.
BUT you should calculate the expected time more accurately. Please calculate the arc length of the odom pose and interpolate the corresponding expected time.
You can use `autoware::motion_utils::findNearestSegmentIndex()`.